### PR TITLE
Add auth

### PR DESCRIPTION
**Pull request by Israel-Laguan**
# Resume

We've created the auth system as described in the project requirements, added signup system as microverse require, and *came to the conclusion that we met all of the requirements*.

# Proposal

We want to have a review of the code to know if we need to make any changes. If no changes needed we want the project to be ACCEPTED so we can move to the next project with a happy smile :)

# Comments
This is really the most challenging of the projects until now in our opinion, including ruby/Enumerable, because of the quantity of different new things to apply in a single project. Challenging but it also give us a lot of input and insight of the Auth system on Rails, so we expect to be using these concepts quite often.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                                  User.digest(User.new_token))
+    cookies.delete(:remember_token)
+    self.current_user = nil
+  end
+
+  attr_writer :current_user
+
+  def current_user
+    remember_token = User.digest(cookies[:remember_token])
+    @current_user ||= User.find_by(remember_token: remember_token)
```
Comment (samgaco) [inline app/controllers/application_controller.rb]: - [ ] The method to retrieve the current user is fine but you left out the method to set the current user as requested in the odin project:
current_user_requirement

File: app/views/posts/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    </tbody>
+</table>
+
+<hr>
+
+<% if @signed_in %>
+<%= link_to 'Sign Out', signout_path, method: :delete %>
+<% else %>
+<%= link_to 'Sign In', new_session_path  %> |
+<%= link_to 'Sign Up', signup_path %>
```
Comment (samgaco) [inline app/views/posts/index.html.erb]: - [ ] This should show in every part of the application and not just only in the post index

sign_in_out_link_requirement

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+class PostsController < ApplicationController
+  before_action :signed_in?, only: %i[new create]
+
+  def index
+    @posts = Post.all
+    @signed_in = current_user.present?
+  end
+
+  def new; end
```
Comment (samgaco) [inline app/controllers/posts_controller.rb]: - [ ] This should be initiating a new post

posts_new_requirement

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (Israel-Laguan) [reply app/controllers/application_controller.rb]: > `attr_writer :current_user`

This line is the one responsible for set the current user. What happened was that rubocop changed the method to this form, and we tested and worked the same. 

If that's not the solution to the requirement, can you provide more explanation on the method we need to provide @samgaco ? Thanks in advance.

Comment (samgaco) [reply app/controllers/application_controller.rb]: Yes, the odin project is asking for a setter. 

Something similar too

```
def current_user(user)
  @current_user = user
end
```

File: app/controllers/posts_controller.rb (hunk 3 above)
Comment (Israel-Laguan) [reply app/controllers/posts_controller.rb]: This empty method goes to the form (in posts/new.html.erb) that creates a post, when you click submit the action is taken on `Posts#create`, successfully creating a new post. If you are referring to add the user that is done in before_action.

If that is not what you are referring at, please Mr. @samgaco provide more information about the requirement and we will be happy to solve the issue.

File: app/views/posts/index.html.erb (hunk 2 above)
Comment (Israel-Laguan) [reply app/views/posts/index.html.erb]: Ok, you are right, we will make this change.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (Israel-Laguan) [reply app/controllers/application_controller.rb]: Yeah, the line that I refer do exactly that. As I told you the linter suggested to change the long method to the one line solution that do the same. Also you can test and do the same result.

Comment (Israel-Laguan) [reply app/controllers/application_controller.rb]: Or are you suggesting that we put the method explicitly not implicitly?

Comment (Israel-Laguan) [reply app/controllers/application_controller.rb]: review
When we try to use it in explicit way, rubocop complains!

So we decided to use as rubocop suggest, also look [this: ](https://medium.com/@gabriellamedas/attr-reader-vs-instance-variables-bdb14249d396)

File: app/views/posts/index.html.erb (hunk 2 above)
Comment (Israel-Laguan) [reply app/views/posts/index.html.erb]: We moved to `views/application.html.erb` and now it show on every part of the application, as you suggested.

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by_email(params[:sessions][:email])
+    p user
```
Comment (Ispirett) [inline app/controllers/sessions_controller.rb]: You can remove the puts user

Comment (Israel-Laguan) [reply app/controllers/sessions_controller.rb]: Yeah, thanks! I'll remove it right now!

**Pull request merged**